test/scripts/e2e_subs/app-accounts.sh

```diff
@@ -0,0 +1,118 @@
+
+my_dir="$(dirname "$0")"
+source "$my_dir/rest.sh" "$@"
+function rest() {
+    curl -q -s -H "Authorization: Bearer $PUB_TOKEN" "$NET$1"
+}
+
+set -e
+set -x
+set -o pipefail
+export SHELLOPTS
+
+WALLET=$1
+
+TEAL=test/scripts/e2e_subs/tealprogs
+
+gcmd="goal -w ${WALLET}"
+
+ACCOUNT=$(${gcmd} account list|awk '{ print $3 }')
+# Create a smaller account so rewards won't change balances.
+SMALL=$(${gcmd} account new | awk '{ print $6 }')
+# Under one algo receives no rewards
+${gcmd} clerk send -a 1000000 -f "$ACCOUNT" -t "$SMALL"
+
+function balance {
+    acct=$1; shift
+    goal account balance -a "$acct" | awk '{print $1}'
+}
+
+[ "$(balance "$ACCOUNT")" = 999998999000 ]
+[ "$(balance "$SMALL")" =        1000000 ]
+
+APPID=$(${gcmd} app create --creator "${SMALL}" --approval-prog=${TEAL}/app-escrow.teal --global-byteslices 4 --global-ints 0 --local-byteslices 0 --local-ints 1  --clear-prog=${TEAL}/approve-all.teal | grep Created | awk '{ print $6 }')
+[ "$(balance "$SMALL")" = 999000 ] # 1000 fee
+
+function appl {
+    method=$1; shift
+    ${gcmd} app call --app-id="$APPID" --from="$SMALL" --app-arg="str:$method" "$@"
+}
+function app-txid {
+    # When app (call or optin) submits, this is how the txid is
+    # printed.  Not in appl() because appl is also used with -o to
+    # create tx
+    grep -o -E 'txid [A-Z0-9]{52}' | cut -c 6- | head -1
+}
+
+APPACCT=$(python -c "import algosdk.encoding as e; print(e.encode_address(e.checksum(b'app'+($APPID).to_bytes(8, 'big'))))")
+
+function payin {
+    amount=$1; shift
+    ${gcmd} clerk send -f "$SMALL" -t "$APPACCT" -a "$amount" "$@"
+}
+
+T=$TEMPDIR
+
+function sign {
+    ${gcmd} clerk sign -i "$T/$1.tx" -o "$T/$1.stx"
+}
+
+TXID=$(${gcmd} app optin --app-id "$APPID" --from "${SMALL}" | app-txid)
+# Rest succeeds, no stray inner-txn array
+[ "$(rest "/v2/transactions/pending/$TXID" | jq '.["inner-txn"]')" == null ]
+[ "$(balance "$SMALL")" = 998000 ] # 1000 fee
+
+appl "deposit():void" -o "$T/deposit.tx"
+payin 150000 -o "$T/pay1.tx"
+cat "$T/deposit.tx" "$T/pay1.tx" | ${gcmd} clerk group -i - -o "$T/group.tx"
+sign group
+${gcmd} clerk rawsend -f "$T/group.stx"
+[ "$(balance "$SMALL")" = 846000 ] # 2 fees, 150,000 deposited
+[ "$(balance "$APPACCT")" = 150000 ]
+
+# Withdraw 20,000 in app. Confirm that inner txn is visible to transaction API.
+TXID=$(appl "withdraw(uint64):void" --app-arg="int:20000" | app-txid)
+[ "$(rest "/v2/transactions/pending/$TXID" \
+        | jq '.["inner-txns"][0].txn.txn.amt')" = 20000 ]
+[ "$(rest "/v2/transactions/pending/$TXID?format=msgpack" | msgpacktool -d \
+        | jq '.["inner-txns"][0].txn.txn.type')" = '"pay"' ]
+# Now confirm it's in blocks API (this time in our internal form)
+ROUND=$(rest "/v2/transactions/pending/$TXID" | jq '.["confirmed-round"]')
+rest "/v2/blocks/$ROUND" | jq .block.txns[0].dt.itx
+
+[ "$(balance "$SMALL")" = 865000 ]   # 1 fee, 20,000 withdrawn
+[ "$(balance "$APPACCT")" = 129000 ] # 20k withdraw, fee paid by app account
```

Comment:
So in this case the fee is "lost"? Meaning the local state for the deposited account still shows 130,000, but the app's balance is only 129,000 now.

This potentially exposes a larger issue: if an app wants to make sure it _never_ pays a fee when creating an inner transaction (e.g. fee credit always used), how could it do that?

Reply:
I guess you could have the app assert that `txn Fee` is always greater than or equal to 2 * `global MinTxnFee`. But due to pooling fees that would be stricter than necessary in some cases (e.g. txn 1 pays a huge fee, txn 2 is an app call with 0 fee but fee credit is still positive).

The way to solve this problem in every situation might be to add a `global FeeCredit` field, which would diminish with each inner txn which draws from the excess fee credit. Though it's not perfect, since inner txn fees can't be inspected, so it would require assuming that their fee will always be `global MinTxnFee`. Then you could assert `global MinTxnFee` > `global FeeCredit` before `tx_submit`-ing.

Not sure if this needs to be addressed at this time, but worth thinking about.

Reply:
Actually it might be better to ask for forgiveness instead of permission. And by that I mean use the `balance` op after `tx_submit` to verify the app account's balance is what you expect (i.e. no fee was taken, if that's the goal). If it's not what you expect, simply reject the app call.